src/lib/protocols/natpmp.c

```diff
@@ -44,70 +47,144 @@ static void ndpi_int_natpmp_add_connection(struct ndpi_detection_module_struct *
+{
+  if (natpmp_is_common_header(packet) == 0)
   {
-    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
-    return;
+    return 0;
   }
 
-  natpmp_type = packet->payload[1];
-  switch (natpmp_type)
+  *natpmp_type = packet->payload[1];
+  switch (*natpmp_type)
   {
     case NATPMP_REQUEST_ADDRESS:
       if (packet->payload_packet_len != 2)
       {
-        NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+        return 0;
       }
-      return;
-
+      break;
     case NATPMP_REQUEST_UDP_MAPPING:
     case NATPMP_REQUEST_TCP_MAPPING:
       if (packet->payload_packet_len != 12 || get_u_int16_t(packet->payload, 2) != 0x0000)
       {
-        NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
-        return;
+        return 0;
       }
       break;
-
-    case NATPMP_RESPONSE_ADRESS:
+    case NATPMP_RESPONSE_ADDRESS:
     case NATPMP_RESPONSE_UDP_MAPPING:
     case NATPMP_RESPONSE_TCP_MAPPING:
-      if ((natpmp_type == NATPMP_RESPONSE_ADRESS && packet->payload_packet_len != 12) ||
-          (natpmp_type != NATPMP_RESPONSE_ADRESS && packet->payload_packet_len != 16))
+      if ((*natpmp_type == NATPMP_RESPONSE_ADDRESS && packet->payload_packet_len != 12) ||
+          (*natpmp_type != NATPMP_RESPONSE_ADDRESS && packet->payload_packet_len != 16))
       {
-        NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
-        return;
+        return 0;
       }
 
       {
         u_int16_t result_code = ntohs(get_u_int16_t(packet->payload, 2));
         if (result_code > 5)
         {
-          NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
-          return;
+          return 0;
         }
       }
       break;
 
     default:
-      NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
-      return;
+      return 0;
+  }
+
+  return 1;
+}
+
+static int ndpi_search_natpmp_extra(struct ndpi_detection_module_struct *ndpi_struct,
+                                    struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct const * const packet = &ndpi_struct->packet;
+  enum natpmp_type natpmp_type;
+
+  if (natpmp_is_valid(packet, &natpmp_type) == 0)
+  {
+    ndpi_set_risk(ndpi_struct, flow, NDPI_MALFORMED_PACKET, "Invalid NATPMP Header");
+    natpmp_disable_extra_dissection(flow);
+    return 0;
+  }
+
+  switch (natpmp_type)
+  {
+    case NATPMP_REQUEST_ADDRESS:
+      return 1; // Nothing to do here.
+    case NATPMP_REQUEST_UDP_MAPPING:
+    case NATPMP_REQUEST_TCP_MAPPING:
+      flow->protos.natpmp.internal_port = ntohs(get_u_int16_t(packet->payload, 4));
+      flow->protos.natpmp.external_port = ntohs(get_u_int16_t(packet->payload, 6));
+      if (flow->protos.natpmp.internal_port == 0)
+      {
+        ndpi_set_risk(ndpi_struct, flow, NDPI_MALFORMED_PACKET, "Request Port Mapping: Internal port must not 0");
+      }
+      break;
+    case NATPMP_RESPONSE_ADDRESS:
+      flow->protos.natpmp.result_code = ntohs(get_u_int16_t(packet->payload, 2));
+      flow->protos.natpmp.external_address.ipv4 = ntohl(get_u_int32_t(packet->payload, 8));
```

Comment:
... and ntohl here... is that right? Unfortunately there is no an example with this parameter

Reply:
It is wrong. `man 3 inet_ntop` tells me that it expects network byte order.